rtengine/simpleprocess.cc

```diff
@@ -1369,41 +1369,75 @@ class ImageProcessor {
             params.crop.enabled = false;
         }
 
-        if (params.resize.enabled) {
-            // resize image
-            LabImage *resized = new LabImage(imw, imh);
-            ipf.Lanczos(tmplab, resized, tmpScale);
-            delete tmplab;
-            tmplab = resized;
-            params.resize.enabled = false;
+        assert(params.resize.enabled);
 
-            params.sharpening = params.prsharpening;
-            params.impulseDenoise.thresh = 
-                int(float(params.impulseDenoise.thresh) * tmpScale);
-            if (tmpScale < 0.5) {
-                params.impulseDenoise.enabled = false;
-            }
-            params.wavelet.strength =
-                int(float(params.wavelet.strength) * tmpScale);
-            params.dirpyrDenoise.luma *= tmpScale;
-            //params.dirpyrDenoise.smethod = "shal";
-
-            const double dirpyreq_scale = min(tmpScale * 1.5, 1.0);
-            for (int i = 0; i < 6; ++i) {
-                auto &n = params.dirpyrequalizer.mult[i];
-                n = 1.0 + (n - 1.0) * dirpyreq_scale;
-            }
+        // resize image
+        LabImage *resized = new LabImage(imw, imh);
+        ipf.Lanczos(tmplab, resized, tmpScale);
+        delete tmplab;
+        tmplab = resized;
+
+        adjust_procparams(tmpScale);
             
-            fw = imw;
-            fh = imh;
-        }
+        fw = imw;
+        fh = imh;
 
         delete baseImg;
         baseImg = new Imagefloat(fw, fh);
         ipf.lab2rgb(*tmplab, *baseImg, params.icm.working);
         delete tmplab;
     }
 
+    void adjust_procparams(double scale_factor)
+    {
+        procparams::ProcParams& params = job->pparams;
+        procparams::ProcParams defaultparams;
+
+#define ADJUST_RADIUS_(n, f)                                              \
```

Comment:
Okay, [generic lambdas](http://stackoverflow.com/a/17233649) (which would be the most obvious solution) are only available in C++14. We could create a function local class to emulate it, but that would lead to more code than before. So we have to use a plain old template function:

```patch
diff --git a/rtengine/simpleprocess.cc b/rtengine/simpleprocess.cc
index fd06df66..d741797c 100644
--- a/rtengine/simpleprocess.cc
+++ b/rtengine/simpleprocess.cc
@@ -1388,23 +1388,24 @@ private:
         delete tmplab;
     }
 
+    template<typename T>
+    void adjustRadius(const T& default_param, double scale_factor, T& param)
+    {
+        const double delta = (param - default_param) * scale_factor;
+        param = default_param + delta;
+    }
+
     void adjust_procparams(double scale_factor)
     {
         procparams::ProcParams& params = job->pparams;
         procparams::ProcParams defaultparams;
 
-#define ADJUST_RADIUS_(n, f)                                              \
-        do {                                                            \
-            auto delta = (params. n - defaultparams. n) * f;            \
-            params. n = defaultparams. n + delta;                       \
-        } while (false)
-        
         params.resize.enabled = false;
 
         if (params.prsharpening.enabled) {
             params.sharpening = params.prsharpening;
         } else {
-            ADJUST_RADIUS_(sharpening.radius, scale_factor);
+            adjustRadius(defaultparams.sharpening.radius, scale_factor, params.sharpening.radius);
         }
         params.impulseDenoise.thresh = 
             int(float(params.impulseDenoise.thresh) * scale_factor);
@@ -1427,13 +1428,13 @@ private:
 
         const double dirpyreq_scale = min(scale_factor * 1.5, 1.0);
         for (int i = 0; i < 6; ++i) {
-            ADJUST_RADIUS_(dirpyrequalizer.mult[i], dirpyreq_scale);
+            adjustRadius(defaultparams.dirpyrequalizer.mult[i], dirpyreq_scale, params.dirpyrequalizer.mult[i]);
             // auto &n = params.dirpyrequalizer.mult[i];
             // n = 1.0 + (n - 1.0) * dirpyreq_scale;
         }
 
-        ADJUST_RADIUS_(defringe.radius, scale_factor);
-        ADJUST_RADIUS_(sh.radius, scale_factor);
+        adjustRadius(defaultparams.defringe.radius, scale_factor, params.defringe.radius);
+        adjustRadius(defaultparams.sh.radius, scale_factor, params.sh.radius);
 
         if (params.raw.xtranssensor.method ==
             procparams::RAWParams::XTransSensor::methodstring[
@@ -1442,8 +1443,6 @@ private:
                 procparams::RAWParams::XTransSensor::methodstring[
                     procparams::RAWParams::XTransSensor::onePass];
         }
-
-#undef ADJUST_RADIUS_
     }
 
 private:
```

HTH,
Flössie